### PR TITLE
Add a reference (HStar2) sparse tree impl for testing.

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: merkle/sparse_merkle_tree_test.go [hunk 1]
```diff
@@ -1,3 +1,157 @@
...
+	"testing"
+
+	"github.com/google/trillian"
+)
+
+// This root was calculated with the C++/Python sparse merkle tree code in the
+// github.com/google/certificate-transparency repo.
+const sparseEmptyRootHashB64 = "xmifEIEqCYCXbZUz2Dh1KCFmFZVn7DUVVxbBQTr1PWo="
+
+func mustDecode(b64 string) trillian.Hash {
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree_test.go]: We should probably start collecting up test helper code in one place. Can leave till later.

File: merkle/sparse_merkle_tree_test.go [hunk 2]
```diff
@@ -1,3 +1,157 @@
...
+		hasher:          h,
+		hStarEmptyCache: []trillian.Hash{h.HashLeaf([]byte(""))},
+	}
+}
+
+func indexLess(a, b *sparseRefValue) bool {
+	return a.index.Cmp(b.index) < 0
+}
+
+func (s *sparseReference) HStar2(n int, values []sparseRefValue) trillian.Hash {
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree_test.go]: Should these ints be int64s or is this good?

File: merkle/sparse_merkle_tree_test.go [hunk 3]
```diff
@@ -1,3 +1,157 @@
...
+	by(indexLess).Sort(values)
+	offset := big.NewInt(0)
+	return s.HStar2b(n, values, 0, len(values), offset)
+}
+
+func (s *sparseReference) HStarEmpty(n int) trillian.Hash {
+	if len(s.hStarEmptyCache) <= n {
+		h := s.hasher.HashChildren(s.HStarEmpty(n-1), s.HStarEmpty(n-1))
+		if len(s.hStarEmptyCache) != n {
+			panic(fmt.Errorf("cache wrong size - expected %d, but cache contains %d entries", n, len(s.hStarEmptyCache)))
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree_test.go]: Pass in testing.T and Fatalf rather than panic? Also below.

File: merkle/sparse_merkle_tree_test.go [hunk 4]
```diff
@@ -1,3 +1,157 @@
...
+
+	split := new(big.Int).Lsh(big.NewInt(1), uint(n-1))
+	split.Add(split, offset)
+	i := lo + sort.Search(hi-lo, func(i int) bool { return values[lo+i].index.Cmp(split) >= 0 })
+	return s.hasher.HashChildren(
+		s.HStar2b(n-1, values, lo, i, offset),
+		s.HStar2b(n-1, values, i, hi, split))
+}
+
+func values(vs map[string]string) []sparseRefValue {
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree_test.go]: values of what?

File: merkle/sparse_merkle_tree_test.go [hunk 5]
```diff
@@ -1,3 +1,157 @@
...
+	value []byte
+}
+
+func (s *sparseRefValue) String() string {
+	return fmt.Sprintf("sparseRefValue{index: %b, value: %v", s.index, s.value)
+}
+
+type by func(a, b *sparseRefValue) bool
+
+func (by by) Sort(values []sparseRefValue) {
```
Comment (Martin2112) [inline merkle/sparse_merkle_tree_test.go]: Dear Mr Pike. Can we please have generic types? Thx bye.

File: merkle/sparse_merkle_tree_test.go (hunk 2 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree_test.go]: I think it's ok for the small number of entries we'll be using in the tests.

File: merkle/sparse_merkle_tree_test.go (hunk 3 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree_test.go]: changed to return an error instead, thought it was a bit cleaner.

File: merkle/sparse_merkle_tree_test.go (hunk 4 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree_test.go]: Renamed, and added some doc for clarity.

File: merkle/sparse_merkle_tree_test.go (hunk 5 above)
Comment (AlCutter) [inline merkle/sparse_merkle_tree_test.go]: `interface{}` is pretty generic :)

Comment (Martin2112) [inline merkle/sparse_merkle_tree_test.go]: I think this is the second copy of this code already though.

**Pull request merged**